EasyCrypt fails to build with Dune 3.7.0 + Merlin

EC builds normally after the standard sequences of `opam` commands, including

```
opam pin alt-ergo 2.4.2
```

Interestingly, before installing `alt-ergo` we were at 5.0.0, but after we downgrade to 4.14.1. `dune` is at 3.7.0.

After

```
opam install merlin
```

EasyCrypt no longer builds

```
File "src/ecUtils.ml", line 633, characters 22-36:
633 |     if s == aout then BatString.copy aout else s
                            ^^^^^^^^^^^^^^
Error: Unbound value BatString.copy
make: *** [build] Error 1 
```

and `opam list` shows (in particular the compiler is bumped back to
5.0.0):

```
alt-ergo              2.4.2       pinned to version 2.4.2
alt-ergo-lib          2.4.2       The Alt-Ergo SMT prover library
alt-ergo-parsers      2.4.2       The Alt-Ergo SMT prover parser library
base-bigarray         base
base-domains          base
base-nnp              base        Naked pointers prohibited in the OCaml heap
base-threads          base
base-unix             base
batteries             3.6.0       A community-maintained standard library extens
camlp-streams         5.0.1       The Stream and Genlex libraries for use with C
camlzip               1.11        Accessing compressed files in ZIP, GZIP and JA
cmdliner              1.1.1       Declarative definition of command line interfa
conf-autoconf         0.1         Virtual package relying on autoconf installati
conf-gmp              4           Virtual package relying on a GMP lib system in
conf-libpcre          1           Virtual package relying on a libpcre system in
conf-pkg-config       2           Check if pkg-config is installed and create an
conf-which            1           Virtual package relying on which
conf-zlib             1           Virtual package relying on zlib
cppo                  1.6.9       Code preprocessor like cpp for OCaml
csexp                 1.5.1       Parsing and printing of S-expressions in Canon
dot-merlin-reader     4.6         Reads config files for merlin
dune                  3.7.0       Fast, portable, and opinionated build system
dune-build-info       3.7.0       Embed build information inside executable
dune-configurator     3.7.0       Helper library for gathering system configurat
dune-private-libs     3.7.0       Private libraries of Dune
dune-site             3.7.0       Embed locations information inside executable 
dyn                   3.7.0       Dynamic type
menhir                20220210    An LR(1) parser generator
menhirLib             20220210    Runtime support library for parsers generated 
menhirSdk             20220210    Compile-time library for auxiliary tools relat
merlin                4.8-500     Editor helper, provides completion, typing and
merlin-lib            4.8-500     Merlin's libraries
num                   1.4         The legacy Num library for arbitrary-precision
ocaml                 5.0.0       The OCaml compiler (virtual package)
ocaml-base-compiler   5.0.0       Official release 5.0.0
ocaml-config          3           OCaml Switch Configuration
ocaml-inifiles        1.2         An ini file parser
ocaml-options-vanilla 1           Ensure that OCaml is compiled with no special 
ocamlbuild            0.14.2      OCamlbuild is a build system with builtin rule
ocamlfind             1.9.6       A library manager for OCaml
ocplib-simplex        0.4         A library implementing a simplex algorithm, in
ordering              3.7.0       Element ordering
pcre                  7.5.0       Bindings to the Perl Compatibility Regular Exp
pp                    1.1.2       Pretty-printing library
psmt2-frontend        0.4.0       The psmt2-frontend project
seq                   base        Compatibility package for OCaml's standard ite
stdlib-shims          0.3.0       Backport some of the new stdlib features to ol
stdune                3.7.0       Dune's unstable standard library
why3                  1.6.0       Why3 environment for deductive program verific
yojson                2.0.2       Yojson is an optimized parsing and printing li
zarith                1.12        Implements arithmetic and logical operations o
```

--- a/README.md
+++ b/README.md
@@ -34,7 +34,7 @@ EasyCrypt uses the following third-party tools/libraries:
 
  * OCamlbuild
 
- * Why3 (= 1.5.x)
+ * Why3 (= 1.6.x)
 
      Available at <http://why3.lri.fr/>
 
@@ -202,15 +202,15 @@ Why3 and SMT solvers are independent pieces of software with their
 own version-specific interactions. Obtaining a working SMT setup may
 require installing specific versions of some of the provers.
 
-At the time of writing, we depend on Why3 1.5.x, which supports the
+At the time of writing, we depend on Why3 1.6.x, which supports the
 following prover versions:
 
- * Alt-Ergo 2.4.1
+ * Alt-Ergo 2.4.2
  * CVC4 1.8
- * Z3 4.8.10
+ * Z3 4.12.1
 
 `alt-ergo` can be installed using opam, if you do you can use pins to
-select a specific version (e.g, `opam pin alt-ergo 2.4.1`). T
+select a specific version (e.g, `opam pin alt-ergo 2.4.1`).
 
 Development branches use `dune-3.x` and which is incompatible with
 `alt-ergo-2.4.1`. In this case, you can use `alt-ergo-2.4.2`. The

--- a/default.nix
+++ b/default.nix
@@ -4,10 +4,10 @@ with import <nixpkgs> {};
 
 let why3_local =
   why3.overrideAttrs (o : rec {
-    version = "1.5.1";
+    version = "1.6.0";
     src = fetchurl {
       url = "https://why3.gitlabpages.inria.fr/releases/${o.pname}-${version}.tar.gz";
-      sha256 = "sha256-vNR7WeiSvg+763GcovoZBFDfncekJMeqNegP4fVw06I=";
+      sha256 = "sha256-hFvM6kHScaCtcHCc6Vezl9CR7BFbiKPoTEh7kj0ZJxw=";
     };
   });
 in

--- a/dune-project
+++ b/dune-project
@@ -21,7 +21,7 @@
   dune-site
   (ocaml-inifiles (>= 1.2))
   (pcre (>= 7))
-  (why3 (and (>= 1.5.0) (< 1.6)))
+  (why3 (and (>= 1.6.0) (< 1.7)))
   yojson
   (zarith (>= 1.10))
 ))

--- a/easycrypt.opam
+++ b/easycrypt.opam
@@ -9,7 +9,7 @@ depends: [
   "dune-site"
   "ocaml-inifiles" {>= "1.2"}
   "pcre" {>= "7"}
-  "why3" {>= "1.5.0" & < "1.6"}
+  "why3" {>= "1.6.0" & < "1.7"}
   "yojson"
   "zarith" {>= "1.10"}
   "odoc" {with-doc}

--- a/examples/MEE-CBC/CBC.eca
+++ b/examples/MEE-CBC/CBC.eca
@@ -30,7 +30,7 @@ op zeros: block.
 op (+): block -> block -> block.
 
 axiom add0b (x : block)    : zeros + x = x.
-axiom addbA (y x z : block): x + (y + z) = (x + y) + z.
+axiom addbA (x y z : block): x + (y + z) = (x + y) + z.
 axiom addbC (x y : block)  : x + y = y + x.
 axiom addbK (x : block)    : x + x = zeros.
 
@@ -38,8 +38,11 @@ clone Ring.ZModule as Block with
   type t               <- block,
   op   zeror           <- zeros,
   op   (+)             <- (+),
-  op   [-] (x : block) <- x
-proof * by smt.
+  op   [-] (x : block) <- x,
+  lemma addrA          <- addbA,
+  lemma addrC          <- addbC,
+  lemma add0r          <- add0b,
+  lemma addNr          <- addbK.
 
 (* -------------------------------------------------------------------- *)
 (** Let P, Pi: key -> block -> block be a PRP and its inverse **)
@@ -514,8 +517,8 @@ section Reduce.
       split=> [|_]; 1:smt (dBlock_uffu).
       split=> [|_]; 1:smt (addbA addbK add0b). (* TODO: instantiate ring structure *)
 split.
-      by move=> h; rewrite -Block.addrA Block.subrr Block.addr0 /= /#.
-      by move=> h; rewrite -Block.addrA Block.subrr Block.addr0 /= /#.
+      by move=> h; rewrite -addbA Block.subrr Block.addr0 /= /#.
+      by move=> h; rewrite -addbA Block.subrr Block.addr0 /= /#.
     wp; rnd (fun x => x + pi{2}).
     auto=> /> &2 szp_neq_0; split=> [|_]; 1:smt (addbA addbK add0b). (* TODO: instantiate ring structure *)
     split=> [|_ s _]; 1:smt (dBlock_uffu).

--- a/examples/MEE-CBC/MAC_then_Pad_then_CBC.eca
+++ b/examples/MEE-CBC/MAC_then_Pad_then_CBC.eca
@@ -43,7 +43,7 @@ op zeros: block.
 op (+): block -> block -> block.
 
 axiom add0b (x : block)    : zeros + x = x.
-axiom addbA (y x z : block): x + (y + z) = (x + y) + z.
+axiom addbA (x y z : block): x + (y + z) = (x + y) + z.
 axiom addbC (x y : block)  : x + y = y + x.
 axiom addbK (x : block)    : x + x = zeros.
 

--- a/shell.nix
+++ b/shell.nix
@@ -1,17 +1,17 @@
 { withProvers ? true, devDeps ? [] }:
 
+with import <nixpkgs> {};
+
 let why3_local =
   why3.overrideAttrs (o : rec {
-    version = "1.5.1";
+    version = "1.6.0";
     src = fetchurl {
       url = "https://why3.gitlabpages.inria.fr/releases/${o.pname}-${version}.tar.gz";
-      sha256 = "sha256-vNR7WeiSvg+763GcovoZBFDfncekJMeqNegP4fVw06I=";
+      sha256 = "sha256-hFvM6kHScaCtcHCc6Vezl9CR7BFbiKPoTEh7kj0ZJxw=";
     };
   });
 in
 let why3 = why3_local; in
-
-with import <nixpkgs> {};
 
 let provers =
   if withProvers then [

--- a/src/ecProvers.ml
+++ b/src/ecProvers.ml
@@ -130,6 +130,8 @@ module Config : sig
 
   val w3_env   : unit -> Env.env
   val provers  : unit -> why3prover list
+  val config   : unit -> Whyconf.config
+  val main     : unit -> Whyconf.main
   val known    : evicted:bool -> prover list
 end = struct
   let theconfig  : (Whyconf.config option) ref = ref None
@@ -149,7 +151,7 @@ end = struct
       let load_prover p config =
         let name    = p.Whyconf.prover_name in
         let version = p.Whyconf.prover_version in
-        let driver  = Whyconf.load_driver_raw main w3_env config.Whyconf.driver [] in
+        let driver  = Driver.load_driver_for_prover main w3_env config in
 
         { pr_prover  =
             { pr_name    = name;
@@ -186,6 +188,12 @@ end = struct
 
   let provers () =
     load (); !theprovers
+
+  let config () =
+    load (); EcUtils.oget !theconfig
+
+  let main () =
+    load (); EcUtils.oget !themain
 
   let known ~evicted =
     let test p =
@@ -356,14 +364,13 @@ let run_prover
         let limit = { Call_provers.empty_limit with
           Call_provers.limit_time =
             let limit = pi.pr_timelimit * pi.pr_cpufactor in
-            if limit <= 0 then 0 else limit;
+            if limit <= 0 then 0. else float_of_int limit;
         } in
 
         let rec doit gcdone =
           try
             Driver.prove_task
-              ~libdir:Why3.Config.libdir
-              ~datadir:Why3.Config.datadir
+              ~config:(Config.main ())
               ~command ~limit dr task
           with Unix.Unix_error (Unix.ENOMEM, "fork", _) when not gcdone ->
             Gc.compact (); doit true
@@ -506,6 +513,6 @@ let execute_task ?(notify : notify option) (pi : prover_infos) task =
         match pcs.(i) with
         | None -> ()
         | Some (_prover, pc) ->
-            CP.interrupt_call ~libdir:Why3.Config.libdir pc;
+            CP.interrupt_call ~config:(Config.main ()) pc;
             (try ignore (CP.wait_on_call pc : CP.prover_result) with _ -> ());
       done)

--- a/src/ecSmt.ml
+++ b/src/ecSmt.ml
@@ -193,37 +193,35 @@ let load_wtheory (genv : tenv) (th : WTheory.theory) : unit =
 (* Create why3 tuple theory with projector                              *)
 
 module Tuples = struct
-
   let ts = Hint.memo 17 (fun n ->
     let vl = ref [] in
-    for _i = 1 to n do
+    for _ = 1 to n do
       vl := WTy.create_tvsymbol (WIdent.id_fresh "a") :: !vl done;
-    WTy.create_tysymbol (WIdent.id_fresh ("tuple" ^ string_of_int n)) !vl WTy.NoDef)
+    WTy.create_tysymbol (WIdent.id_fresh (Format.sprintf "tuple%d" n)) !vl WTy.NoDef)
 
   let proj = Hdint.memo 17 (fun (n, k) ->
-    if not (0 <= k && k < n) then (Format.eprintf " k = %i; n = %i@." k n; assert false);
+    assert (0 <= k && k < n);
     let ts = ts n in
     let tl = List.map WTy.ty_var ts.WTy.ts_args in
     let ta = WTy.ty_app ts tl in
     let tr = List.nth tl k in
-    let id =
-      WIdent.id_fresh ("proj" ^ string_of_int n ^ "_" ^ string_of_int k) in
-    WTerm.create_fsymbol id [ta] tr)
+    let id = WIdent.id_fresh (Format.sprintf "proj%d_%d" n k) in
+    WTerm.create_fsymbol ~proj:true id [ta] tr)
 
   let fs = Hint.memo 17 (fun n ->
     let ts = ts n in
     let tl = List.map WTy.ty_var ts.WTy.ts_args in
     let ty = WTy.ty_app ts tl in
-    let id = WIdent.id_fresh ("Tuple" ^ string_of_int n) in
+    let id = WIdent.id_fresh (Format.sprintf "Tuple%d" n) in
     WTerm.create_fsymbol ~constr:1 id tl ty)
 
   let theory = Hint.memo 17 (fun n ->
     let ts = ts n and fs = fs n in
     let pl = List.mapi (fun i _ -> Some (proj (n, i))) ts.WTy.ts_args in
     let uc =
-      WTheory.create_theory ~path:["Easycrypt"]
-        (WIdent.id_fresh ("Tuple" ^ string_of_int n))  in
-    let uc = WTheory.add_data_decl uc [ts, [fs,pl]] in
+      let name = Format.sprintf "Tuple%d" n in
+      WTheory.create_theory ~path:["Easycrypt"] (WIdent.id_fresh name)  in
+    let uc = WTheory.add_data_decl uc [ts, [fs, pl]] in
     WTheory.close_theory uc)
 
 end
@@ -450,7 +448,7 @@ and trans_tydecl genv (p, tydecl) =
         let for_field (fname, fty) =
           let wfid  = pqoname (prefix p) fname in
           let wfty  = trans_ty (genv, lenv) fty in
-          let wcls  = WTerm.create_lsymbol (preid_p wfid) [wdom] (Some wfty) in
+          let wcls  = WTerm.create_lsymbol ~proj:true (preid_p wfid) [wdom] (Some wfty) in
           let w3op  = plain_w3op ~name:fname tparams wcls in
           ((wfid, w3op), wcls)
         in

--- a/src/ecUtils.ml
+++ b/src/ecUtils.ml
@@ -628,12 +628,6 @@ module String = struct
 
   let split_lines = split_on_string ~by:"\n"
 
-  let trim (s : string) =
-    let aout = BatString.trim s in
-    if s == aout then BatString.copy aout else s
-
-  let rev (s:string) = init (length s) (fun i -> s.[length s - 1 - i])
-
   (* ------------------------------------------------------------------ *)
   module OptionMatching = struct
     let all_matching tomatch s =

--- a/theories/algebra/StdBigop.ec
+++ b/theories/algebra/StdBigop.ec
@@ -143,7 +143,7 @@ clone include Bigalg.BigOrder with
     op Num.minr    = fun (x y : real), if x <= y then x else y,
     op Num.maxr    = fun (x y : real), if y <= x then x else y
 
-    proof Num.Domain.* by smt, Num.Axioms.* by smt
+    proof Num.Domain.* by smt(RField.invr0), Num.Axioms.* by smt()
 
     rename [theory] "BAdd" as "BRA"
            [theory] "BMul" as "BRM"
@@ -234,7 +234,7 @@ clone Bigop as BBA with
   type t <- bool,
   op Support.idm <- false,
   op Support.(+) <- Bool.( ^^ )
-  proof Support.Axioms.* by (delta; smt).
+  proof Support.Axioms.* by (delta; smt()).
 
 (* -------------------------------------------------------------------- *)
 theory BBM.
@@ -242,7 +242,7 @@ clone include Bigop with
   type t <- bool,
   op Support.idm <- true,
   op Support.(+) <- Pervasive.( /\ )
-  proof Support.Axioms.* by (delta; smt).
+  proof Support.Axioms.* by (delta; smt()).
 
 lemma bigP (P : 'a -> bool) (s : 'a list):
   big predT P s <=> (forall a, mem s a => P a).
@@ -259,7 +259,7 @@ clone include Bigop with
   type t <- bool,
   op Support.idm <- false,
   op Support.(+) <- Pervasive.( || )
-  proof Support.Axioms.* by (delta; smt).
+  proof Support.Axioms.* by (delta; smt()).
 
 lemma bigP (P : 'a -> bool) (s : 'a list):
   big predT P s <=> (exists a, mem s a /\ P a).

--- a/theories/algebra/StdOrder.ec
+++ b/theories/algebra/StdOrder.ec
@@ -26,7 +26,7 @@ clone include Number.RealDomain
   op   ( <= ) <- Int.(<=),
   op   ( <  ) <- Int.(< )
 
-  proof Domain.* by smt, Axioms.* by smt
+  proof Domain.* by smt(invr0), Axioms.* by smt()
 
   remove abbrev Domain.(-)
   remove abbrev Domain.(/).
@@ -96,7 +96,7 @@ clone include Number.RealField
   op   ( <= ) <- Real.(<=),
   op   ( <  ) <- Real.(< )
 
-  proof Field.* by smt, Axioms.* by smt full
+  proof Field.* by smt(RField.invr0), Axioms.* by smt()
 
   remove abbrev Field.(-)
   remove abbrev Field.(/).

--- a/theories/crypto/prp_prf/Strong_RP_RF.eca
+++ b/theories/crypto/prp_prf/Strong_RP_RF.eca
@@ -141,10 +141,10 @@ qed.
 phoare Indirect_ll: [Indirect.sample: exists x, support uD x /\ !X x ==> true] = 1%r.
 proof.
 proc; seq  1: (exists x, support uD x /\ !X x)=> //=.
-+ by rnd (predT); skip; smt ml=0 w=uD_uf_fu.
++ by rnd (predT); skip; smt(uD_uf_fu).
 if=> //=.
 + rnd (predT); skip.
-  by progress [-split]; split=> //=; smt.
+  progress [-split]; split=> //=; smt.
 by hoare; rnd=> //=; skip=> &hr ->.
 qed.
 
@@ -153,7 +153,7 @@ proof.
 proc; if=> //=; auto; call Indirect_ll.
 skip=> /> &hr x_notin_m.
 have [x0] x0_notinr_m := endo_dom_rng PRPi.m{hr} _; first by exists x{hr}.
-by exists x0; rewrite x0_notinr_m /=; smt w=uD_uf_fu.
+by exists x0; rewrite x0_notinr_m /=; smt(uD_uf_fu).
 qed.
 
 lemma PRPi'_Indirect_fi_ll: islossless PRPi'(Indirect).fi.
@@ -161,7 +161,7 @@ proof.
 proc; if=> //=; auto; call Indirect_ll.
 skip=> /> &hr x_notin_mi.
 have [x0] x0_notinr_mi := endo_dom_rng PRPi.mi{hr} _; first by exists x{hr}.
-by exists x0; rewrite x0_notinr_mi; smt w=uD_uf_fu.
+by exists x0; rewrite x0_notinr_mi; smt(uD_uf_fu).
 qed.
 
 (** The proof is cut into 3 parts (sections):
@@ -237,7 +237,7 @@ section Upto.
     <= Pr[IND(ARP,D).main() @ &m: ARP.coll].
   proof.
   rewrite (PRPi'_Indirect_eq &m).
-  byequiv: PRP_indirect_bad.bad=> //=; 2:smt ml=0.
+  byequiv: PRP_indirect_bad.bad=> //=; 2:smt().
   proc.
   call (_: ARP.coll,
              !PRP_indirect_bad.bad{1} /\ ={m,mi}(PRPi,ARP),
@@ -253,7 +253,8 @@ section Upto.
       by auto=> /#.
     sp; if{1}.
       conseq (_: PRP_indirect_bad.bad{1} /\ ARP.coll{2})=> //=.
-      auto; progress [-split]; split=> //= [|_]; 1:smt.
+      auto; progress [-split]; split=> //= [|_].
+        by apply/excepted_lossless => /#.
       by progress; right.
     by auto; progress [-split]; rewrite H0 /= => -> /=; rewrite get_setE.
   + move=> &2 bad; conseq (_: true ==> true: =1%r)
@@ -261,11 +262,11 @@ section Upto.
       by proc; if=> //=; inline *; seq  2: PRP_indirect_bad.bad; [auto|if=> //=; auto].
     proc; if=> //=; inline *.
     seq  2: (X = rng PRPi.m /\ x \notin PRPi.m) 1%r 1%r 0%r _ => //=.
-      by auto; rewrite -/predT; smt ml=0 w=uD_uf_fu. (* predT should be an abbreviation *)
-      by if=> //=; auto; smt.
+      by auto; rewrite -/predT; smt(uD_uf_fu).
+      by if=> //=; auto=> /> *; apply/excepted_lossless => /#.
       by hoare; auto.
   + move=> &1.
-    proc; if; auto; progress [-split]; rewrite -/predT; split=> //= [|_]; 1:smt ml=0 w=uD_uf_fu.
+    proc; if; auto; progress [-split]; rewrite -/predT; split=> //= [|_]; 1:smt(uD_uf_fu).
     by progress [-split]; rewrite H.
   + proc. if=> //=; inline *.
     swap{1} 1.
@@ -277,7 +278,8 @@ section Upto.
       by auto=> /#.
     sp; if{1}.
       conseq (_: PRP_indirect_bad.bad{1} /\ ARP.coll{2})=> //=.
-      auto; progress [-split]; split=> //= [|_]; 1:smt.
+      auto; progress [-split]; split=> //= [|_].
+        by apply/excepted_lossless => /#.
       by progress; right.
     by auto; progress [-split]; rewrite H0 => /= ->; rewrite get_setE.
   + move=> &2 bad; conseq (_: true ==> true: =1%r)
@@ -285,13 +287,13 @@ section Upto.
       by proc; if=> //=; inline *; seq  2: PRP_indirect_bad.bad; [auto|if=> //=; auto].
     proc; if=> //=; inline *.
     seq  2: (X = rng PRPi.mi /\ y \notin PRPi.mi) 1%r 1%r 0%r _ => //=.
-      by auto; rewrite -/predT; smt ml=0 w=uD_uf_fu. (* predT should be an abbreviation *)
-      by if=> //=; auto; smt.
+      by auto; rewrite -/predT; smt(uD_uf_fu).
+      by if=> //=; auto=> /> *; apply/excepted_lossless => /#.
       by hoare; auto.
   + move=> &1.
-    proc; if; auto; progress [-split]; rewrite -/predT; split=> //= [|_]; 1:smt ml=0 w=uD_uf_fu.
+    proc; if; auto; progress [-split]; rewrite -/predT; split=> //= [|_]; 1:smt(uD_uf_fu).
     by progress [-split]; rewrite H.
-  by inline *; auto; progress; smt.
+  by inline *; auto; progress => /#.
   qed.
 end section Upto.
 
@@ -408,13 +410,13 @@ section CollisionProbability.
             apply/subset_leq_fcard=> x'; rewrite !inE !mem_frng !rngE=> - /= [x0].
             rewrite get_setE; case: (x0 = x{2})=> [<<- /= ->|_ mx0] //.
             by left; exists x0.
-          smt.
+          by rewrite fcardU fcard1; smt(fcard_ge0).
         + apply/(ler_trans (card (frng ARP.mi{2} `|` fset1 x{2}))).
             apply/subset_leq_fcard=> x'; rewrite !inE !mem_frng !rngE=> - /= [x0].
             rewrite get_setE; case: (x0 = yL)=> [<<- //|_ mx0].
             by left; exists x0.
-          smt.
-      - smt ml=0.
+          by rewrite fcardU fcard1; smt(fcard_ge0).
+      - smt().
     * proc; sp; if=> //=. rcondt{2} 1; first by auto=> /#.
       inline *; sp; if=> //=; auto.
       - progress.
@@ -422,13 +424,13 @@ section CollisionProbability.
             apply/subset_leq_fcard=> x'; rewrite !inE !mem_frng !rngE=> - /= [x0].
             rewrite get_setE; case: (x0 = x0L)=> [<<- //|_ mx0].
             by left; exists x0.
-          smt.
+          by rewrite fcardU fcard1; smt(fcard_ge0).
         + apply/(ler_trans (card (frng ARP.mi{2} `|` fset1 x0L))).
             apply/subset_leq_fcard=> x'; rewrite !inE !mem_frng !rngE=> - /= [x0].
             rewrite get_setE; case: (x0 = x{2})=> [<<- //|_ mx0].
             by left; exists x0.
-          smt.
-      - smt ml=0.
+          by rewrite fcardU fcard1; smt(fcard_ge0).
+      - smt().
     by auto=> />; rewrite frng0 fcards0.
   fel 2 FEL.c (fun x, x%r * mu uD (pred1 witness)) q (ARP.coll) 
            [FEL(D).FBounder.f: (FEL.c < q); FEL(D).FBounder.fi: (FEL.c < q)]
@@ -441,7 +443,7 @@ section CollisionProbability.
     proc; sp; rcondt 1=> //.
     inline *; sp; if=> //=; last first.
     * hoare; auto=> // /> &hr 6?.
-      by apply/RealOrder.mulr_ge0; smt w=(mu_bounded ge0_q).
+      by apply/RealOrder.mulr_ge0; smt(mu_bounded ge0_q).
     sp; if=> //=.
     * wp; rnd (rng ARP.m); skip.
       progress.
@@ -450,9 +452,9 @@ section CollisionProbability.
         + by apply/fun_ext=> x; rewrite mem_frng.
         apply/mu_mem_le; move=> x _; have [] uD_suf [] ? uD_fu:= uD_uf_fu.
         apply/RealOrder.lerr_eq/uD_suf; 1,2:rewrite uD_fu //.
-        by apply/RealOrder.ler_wpmul2r; smt w=(mu_bounded lt_fromint ltrW leq_card_rng_dom).
+        by apply/RealOrder.ler_wpmul2r; smt(mu_bounded lt_fromint ltrW leq_card_rng_dom).
       - by move: H9;rewrite H1.
-    * by hoare; auto=> //=; smt w=(RealOrder.mulr_ge0 mu_bounded ge0_q).
+    * by hoare; auto=> //=; smt(RealOrder.mulr_ge0 mu_bounded ge0_q).
   + move=> c; proc. rcondt 2; 1:by auto.
     sp; if=> //=.
     * inline*;sp;if;auto; 2:smt().
@@ -463,7 +465,7 @@ section CollisionProbability.
   + exists*FEL.c;elim*=> c.
     conseq(:_==>_ : (c%r * mu1 uD witness));progress.
     proc; sp; rcondt 1=> //=.
-    inline *; sp; if=> //=; last by hoare; auto; smt w=(RealOrder.mulr_ge0 mu_bounded ge0_q).
+    inline *; sp; if=> //=; last by hoare; auto; smt(RealOrder.mulr_ge0 mu_bounded ge0_q).
     sp; if=> //=.
     * wp; rnd (rng ARP.mi); skip.
       progress.
@@ -472,9 +474,9 @@ section CollisionProbability.
         + by apply/fun_ext=> x; rewrite mem_frng.
         apply/mu_mem_le; move=> x _; have [] uD_suf [] _ uD_fu:= uD_uf_fu.
         apply/RealOrder.lerr_eq/uD_suf; 1,2:rewrite uD_fu //.
-        smt w=(RealOrder.ler_wpmul2r mu_bounded le_fromint ltrW leq_card_rng_dom).
+        smt(RealOrder.ler_wpmul2r mu_bounded le_fromint ltrW leq_card_rng_dom).
       - by move: H9; rewrite H1.
-    * by hoare; auto; smt w=(RealOrder.mulr_ge0 mu_bounded ge0_q).
+    * by hoare; auto; smt(RealOrder.mulr_ge0 mu_bounded ge0_q).
   + move=> c; proc; rcondt 2; 1:by auto.
     sp; if=> //=.
     * inline*;sp;if;auto; 2:smt().

